File: Graph/UnionFind.java
public class UnionFind {
    int [] parents;
    public UnionFind(int n) {
        parents = new int[n];
        for (int i = 0; i < n; i++)  {
            parents[i] = -1;
        }
    }

    int find(int p){
        // find parents of p
        int parent = p;
        while(parents[parent] > 0){
            parent = parents[parent];
        }
        return parent;
    }

    void connect(int p, int q){
        int pParent = find(p);
        int qParent = find(q);
        if(parents[pParent] < parents[qParent]){
            // p parent is larger
            parents[pParent] += parents[qParent];
            parents[qParent] = pParent;
            parents[q] = pParent;
        }else{
            // p parent is larger
            parents[qParent] += parents[pParent];
            parents[pParent] = qParent;
            parents[p] = qParent;
        }
    }

    boolean isConnected(int p, int q){
        int pParent = find(p);
        int qParent = find(q);
        return pParent == qParent;
    }

}

File: algorithm_template/unionfind.py
class unionFind:
    def __init__(self, n) -> None:
        self.parent = [i for i in range(n)]
        self.rank = [1]*n

    def union(self, x, y):
        root_x = self.find(x)
        root_y = self.find(y)
        if root_x == root_y:
            return

        if self.rank[root_x] < self.rank[root_y]:
            self.parent[root_x] = root_y
            self.rank[root_y] += self.rank[root_x]
        else:
            self.parent[root_y] = root_x
            self.rank[root_x] += self.rank[root_y]

    def find(self, x):
        # find the highest parent of x
        if self.parent[x] != x:
            self.parent[x] = self.find(self.parent[x])
        return self.parent[x]

    def isConnected(self, x, y):

        return self.find(x) == self.find(y)
